### PR TITLE
Add state dict hooks

### PR DESCRIPTION
**Pull request by cli99**
This PR adds
1.  `split_expert_weights` to control whether the HF checkpoint is loaded into `DbrxExpertGLU`(with fused expert weights) or `DbrxExpertGLUSplit` (with split expert weights and defined as sequence of nn.linear)
2. `fuse_expert_weights_on_save` to control wether the model state_dict is saved as fused expert weights or not
We [register a load_state_dict_pre_hook ](https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L421C5-L421C31)to `DbrxForCausalLM` to do the `state_dict` remapping. The hook is called before loading `state_dict` into `self`. 

Example usage:
```
        config = AutoConfig.from_pretrained(args.name_or_path,
                                            **from_pretrained_kwargs)
       #  config.ffn_config.split_expert_weights == True to load fused expert weights into split nn.linears
       #  config.ffn_config.fuse_expert_weights_on_save == True if saving model state dict back into fused expert weight
        model = AutoModelForCausalLM.from_pretrained(args.name_or_path,
                                                     config=config,
                                                     torch_dtype=model_dtype,
                                                     device_map='auto',
                                                     low_cpu_mem_usage=False,
                                                     **from_pretrained_kwargs)
```
Changes in modeling_utils.py:
1. For `model._load_state_dict` and `load_state_dict_pre_hook` on the model to be invoked, we reply on [this path](https://github.com/huggingface/transformers/blob/4d8427f/src/transformers/modeling_utils.py#L4008) and need to use `low_cpu_mem_usage`. We don't overwrite `low_cpu_mem_usage` when `device_map == 'auto'`, otherwise [the other code path](https://github.com/huggingface/transformers/blob/4d8427f/src/transformers/modeling_utils.py#L3981) is taken and does not trigger the `load_state_dict_pre_hook`
2. [This](https://github.com/huggingface/transformers/blob/4d8427f/src/transformers/modeling_utils.py#L3838) causes hanging if we configure the model with `split_expert_weights`. So we bypass it if using `split_expert_weights`.

### CODE REVIEW

**Comment by abhi-mosaic** [comment]
>[This](https://github.com/huggingface/transformers/blob/4d8427f/src/transformers/modeling_utils.py#L3838) causes hanging if we configure the model with split_expert_weights. So we bypass it if using split_expert_weights.

@cli99 could you clarify what you mean by a hang here? I think this code path eventually calls `model._init_weights()` which should be overriden/owned by our `DbrxPretrainedModel(PretrainedModel)` class: https://github.com/huggingface/transformers/blob/517a3e670d8fc11374895e870dd0dd041467c7fe/src/transformers/modeling_utils.py#L1689-L1696

Is it possible that it's just running really slow? Perhaps we can use the `no_init_weights` context manager to check: https://github.com/huggingface/transformers/blob/517a3e670d8fc11374895e870dd0dd041467c7fe/src/transformers/modeling_utils.py#L179

Basically i'm trying to see if we can avoid modifying `modeling_utils.py` as required by (2.) in your comment

**Comment by michaelroyzen** [comment]
I'm having trouble using this branch with DeepSpeed ZeRO 3 when starting a training run @abhi-mosaic @cli99. I get the following traceback:

```
TypeError: 'ClassInstantier' object is not callable

  File "/home/ubuntu/.local/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 503, in 
wrapper
    self.blocks = nn.ModuleList([DbrxBlock(config, block_idx) for block_idx in range(config.n_layers)])
  File "/home/ubuntu/.local/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 503, in 
wrapper
    self.experts = DbrxExperts(
  File "/home/ubuntu/.local/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 503, in 
wrapper
    f(module, *args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 894, in __ini
t__
    self.activation_fn = ACT2FN(act_fn_name)
TypeError: 'ClassInstantier' object is not callable
    self.ffn = DbrxFFN(config=config)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 503, in 
wrapper
    f(module, *args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 820, in __ini
t__
    f(module, *args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 870, in __ini
t__
    f(module, *args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 805, in __ini
t__
    self.experts = DbrxExperts(
  File "/home/ubuntu/.local/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 503, in 
wrapper
    f(module, *args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 820, in __ini
t__
    f(module, *args, **kwargs)
    self.mlp = DbrxExpertGLUSplit(hidden_size=hidden_size,
  File "/home/ubuntu/.local/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 503, in 
wrapper
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 820, in __ini
t__
    self.activation_fn = ACT2FN(act_fn_name)
TypeError: 'ClassInstantier' object is not callable
    self.mlp = DbrxExpertGLUSplit(hidden_size=hidden_size,
  File "/home/ubuntu/.local/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 503, in 
wrapper
    self.mlp = DbrxExpertGLUSplit(hidden_size=hidden_size,
  File "/home/ubuntu/.local/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 503, in 
wrapper
    f(module, *args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 805, in __ini
t__
    f(module, *args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 805, in __ini
t__
    self.activation_fn = ACT2FN(act_fn_name)
    f(module, *args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 805, in __ini
t__
TypeError: 'ClassInstantier' object is not callable
    self.activation_fn = ACT2FN(act_fn_name)
TypeError: 'ClassInstantier' object is not callable
    self.activation_fn = ACT2FN(act_fn_name)
TypeError: 'ClassInstantier' object is not callable
```

I'm calling the model like so:

```
config = transformers.AutoConfig.from_pretrained(model_args.model_name_or_path)
config.ffn_config.split_expert_weights = True
config.ffn_config.fuse_expert_weights_on_save = True

model = transformers.AutoModelForCausalLM.from_pretrained(
    model_args.model_name_or_path,
    config=config,
    cache_dir=training_args.cache_dir,
    torch_dtype=torch.bfloat16,
    attn_implementation="flash_attention_2"
)
```

I also replaced the typo `self.is_casual` with `self.is_causal` at https://github.com/abhi-mosaic/transformers/blob/0a9d48c80154d55b42363e4998443d527a9e273f/src/transformers/models/dbrx/modeling_dbrx.py#L255.

**Comment by russellsayshi** [comment]
I found the issue with
```
TypeError: 'ClassInstantier' object is not callable
    self.activation_fn = ACT2FN(act_fn_name)
```

Line 805 in `transformers/models/dbrx/modeling_dbrx.py` should be changed from:
```
self.activation_fn = ACT2FN(act_fn_name)
```
to
```
self.activation_fn = ACT2FN[act_fn_name]
```

However, after applying this fix, I am seeing a new issue crop up:
```
Traceback (most recent call last):                                                                                                                                                                      
  File "/mnt/nfsfast/training/dbrx/train.py", line 270, in <module>                                                                                                                                     
    hook_result = hook(self, destination, prefix, local_metadata)                                                                                                                                       
  File "/home/ubuntu/.local/lib/python3.10/site-packages/transformers/models/dbrx/modeling_dbrx.py", line 1325, in _state_dict_hook                                                                     
    train()                                                                                                                                                                                             
  File "/mnt/nfsfast/training/dbrx/train.py", line 220, in train                                                                                                                                        
    ws = ws.view(moe_num_experts * ffn_hidden_size, hidden_size)                                                                                                                                        
RuntimeError: shape '[172032, 6144]' is invalid for input of size 0                                                                                                                                     
```
It's trying to `view` the `ws` tensor, but `ws` is of size `torch.Size([16, 0])`

Would anyone have any advice for how to resolve this issue? Thank you very much

**Comment by russellsayshi** [comment]
Looks like the issue is happening during the loading of the state dict here

```
        def _state_dict_hook(module, state_dict, prefix, local_metadata):
            """Returns the state dict of the model."""
            n_layers = self.config.n_layers
            moe_num_experts = self.config.ffn_config.moe_num_experts
            ffn_hidden_size = self.config.ffn_config.ffn_hidden_size
            hidden_size = self.config.d_model

            for idx in range(n_layers):
                for w in ["w1", "v1", "w2"]:
                    ws = []
                    key =  f'transformer.blocks.{idx}.ffn.experts.mlp.{w}'
                    new_key = f'transformer.blocks.{idx}.ffn.experts.mlp.{w}'

                    for expert_idx in range(moe_num_experts):
                        inner_key = key + f'.{expert_idx}.weight'
                        if inner_key not in state_dict:
                            inner_key = '_fsdp_wrapped_module.' + inner_key
                        if inner_key not in state_dict:
                            continue
                        ws.append(state_dict[inner_key])

                    if len(ws) > 0 and isinstance(ws[0], torch.Tensor):
                        ws = torch.stack(ws)
                        ws = ws.view(moe_num_experts * ffn_hidden_size, hidden_size)
                        state_dict[new_key] = ws

                        # Remove old keys
                        for expert_idx in range(moe_num_experts):
                            inner_key = key + f'.{expert_idx}.weight'
                            if inner_key not in state_dict:
                                inner_key = '_fsdp_wrapped_module.' + inner_key
                            del state_dict[inner_key]
```

Is it intentional that `key == new_key`?
```
                    key =  f'transformer.blocks.{idx}.ffn.experts.mlp.{w}'
                    new_key = f'transformer.blocks.{idx}.ffn.experts.mlp.{w}'
```

**Comment by michaelroyzen** [comment]
Would love to hear your thoughts @cli99

**Comment by Rocketknight1** [comment]
Hi all, Matt from Hugging Face here. I can't reproduce the issue here but it's possible that the `_state_dict_hook` is being called during Zero offloading, when the tensors haven't yet been fully constructed, and not just when the state dict is being saved. One possible workaround would be to replace the conditional with `if len(ws) > 0 and isinstance(ws[0], torch.Tensor) and torch.numel(ws[0]) > 0`. This would ensure that it is only called for properly built tensors.

That said, though, it's just a guess since I can't reproduce the issue!

**Comment by michaelroyzen** [comment]
Thanks @Rocketknight1. We were using ZeRO offloading, but we get the same crash even without it. 

May I ask how you were able to get the code to run? There seem to be some TypeErrors in this PR that prevent the code from running in the first place. Could you share your code if possible?

**Comment by Rocketknight1** [comment]
@michaelroyzen I basically followed your code and it worked for me! I made a random init of the model with 2 layers, just to make it easy to work with on my local machine, but the code below ran fine using this PR branch, with the edits recommended above (`ACT2FN(` -> `ACT2FN[` and `self.is_casual` -> `self.is_causal`):
```python
>>> from transformers import AutoModelForCausalLM, AutoConfig
>>> config = AutoConfig.from_pretrained("databricks/dbrx-instruct")
>>> config.n_layers = 2
>>> config.ffn_config.split_expert_weights = True
>>> config.ffn_config.fuse_expert_weights_on_save = True
>>> model = AutoModelForCausalLM.from_config(config)
>>> model.save_pretrained("test_dbrx_save")
>>> del model
>>> model = AutoModelForCausalLM.from_pretrained("./test_dbrx_save")
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████| 8/8 [00:12<00:00,  1.57s/it]
>>> model(**model.dummy_inputs)  # This works fine!
```

**Comment by cli99** [comment]
Thank you all for the feedbacks. We are closing this and adding a megablocks codepath in dbrx
